README.md
# Endmondo Sport Tracker Connector

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

Endomondo sport tracker connector.

## Install

Via Composer

``` bash
$ composer require dragosprotung/stc-endomondo
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email dragos@protung.ro instead of using the issue tracker.

## Credits

- [Dragos Protung][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/dragosprotung/stc-endomondo.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/dragosprotung/stc-endomondo/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/dragosprotung/stc-endomondo.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/dragosprotung/stc-endomondo.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/dragosprotung/stc-endomondo.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/dragosprotung/stc-endomondo
[link-travis]: https://travis-ci.org/dragosprotung/stc-endomondo
[link-scrutinizer]: https://scrutinizer-ci.com/g/dragosprotung/stc-endomondo/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/dragosprotung/stc-endomondo
[link-downloads]: https://packagist.org/packages/dragosprotung/stc-endomondo
[link-author]: https://github.com/dragosprotung
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `Sport Tracker Connector Endomondo` will be documented in this file

## 1.0 - 2016-xx-xx

### Added
- Initial split from the sport tracker connector application


tests/API/Workouts/WorkoutsTest.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo\Test\API\Workouts;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;
use Psr\Http\Message\RequestInterface;
use SportTrackerConnector\Core\Workout\Track;
use SportTrackerConnector\Core\Workout\TrackPoint;
use SportTrackerConnector\Core\Workout\Workout;
use SportTrackerConnector\Endomondo\API\Authentication;
use SportTrackerConnector\Endomondo\API\Exception\BadResponseException;
use SportTrackerConnector\Endomondo\API\Workouts;
use SportTrackerConnector\Endomondo\SportMapper;

/**
 * Test case for Workouts.
 */
class WorkoutsTest extends \PHPUnit_Framework_TestCase
{
    public function testGetWorkoutThrowsExceptionOnInvalidAuthentication()
    {
        $authentication = Authentication::fromToken('wrong_token');

        $mockHandler = new MockHandler(
            [
                new Response(200, [], '{"error":{"type":"AUTH_FAILED"}}')
            ]
        );

        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);

        $this->expectException(BadResponseException::class);

        $workouts = new Workouts($authentication, $client);
        $workouts->getWorkout('123456');
    }

    public function testGetWorkoutSuccess()
    {
        $authentication = Authentication::fromToken('token');

        $responseFixtureFile = __DIR__ . '/Fixtures/' . $this->getName() . '.json';
        $mockHandler = new MockHandler(
            [
                new Response(200, [], file_get_contents($responseFixtureFile))
            ]
        );

        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);

        $workouts = new Workouts($authentication, $client);
        $actual = $workouts->getWorkout('123456');
        static::assertJsonStringEqualsJsonFile($responseFixtureFile, json_encode($actual));
    }

    public function testListWorkoutsWithNoMoreData()
    {
        $authentication = Authentication::fromToken('token');

        $mockHandler = new MockHandler(
            [
                new Response(200, [], file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.json'))
            ]
        );
        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);

        $workouts = new Workouts($authentication, $client);
        $actual = $workouts->listWorkouts(new \DateTime('-2 weeks'), new \DateTime('now'));
        static::assertJsonStringEqualsJsonFile(
            __DIR__ . '/Expected/' . $this->getName() . '.json',
            json_encode($actual)
        );
    }

    public function testPostTrackSuccess()
    {
        $trackPoints = array_merge($this->getTrackPointMocks(150, true), $this->getTrackPointMocks(140));

        $track = new Track($trackPoints);

        $workout = $this->createMock(Workout::class);
        $workout->expects(static::any())->method('tracks')->will(static::returnValue(array($track)));

        $authentication = Authentication::fromToken('token');

        $container = array();
        $history = Middleware::history($container);

        $mockHandler = new MockHandler(
            [
                new Response(200, [], file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.txt')),
                new Response(200, [], file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.txt')),
                new Response(200, [], file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.txt')),
                new Response(200, [], file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.txt'))
            ]
        );
        $handler = HandlerStack::create($mockHandler);
        $handler->push($history);

        $client = new Client(['handler' => $handler]);

        $workouts = new Workouts($authentication, $client);
        $workoutId = $workouts->postTrack($track, SportMapper::RUNNING);

        foreach ($container as $i => $transaction) {
            /** @var RequestInterface $request */
            $request = $transaction['request'];
            static::assertSame('POST', $request->getMethod());

            static::assertStringEqualsFile(
                __DIR__ . '/Expected/' . $this->getName() . '-' . $i . '.txt',
                gzdecode($request->getBody()->getContents())
            );
        }

        static::assertSame('123456789', $workoutId);
    }

    /**
     * Get a number of track point mocks.
     *
     * @param integer $number Number of mocks to get.
     * @param boolean $distance Flag if distance should be present in the track point.
     * @return \SportTrackerConnector\Core\Workout\TrackPoint[]
     */
    private function getTrackPointMocks(int $number, bool $distance = false)
    {
        $mocks = array();
        for ($i = 0; $i < $number; $i++) {
            $mocks[] = $this->getTrackPointMock($distance ? $i : null);
        }

        return $mocks;
    }

    /**
     * Get a track point mock.
     *
     * @param integer $distance The distance for the point.
     * @return \SportTrackerConnector\Core\Workout\TrackPoint
     */
    private function getTrackPointMock($distance = null)
    {
        static $call = 1;

        $dateTime = new \DateTime('2016-01-01 00:00:00');
        $dateTime->add(\DateInterval::createFromDateString('+' . $call . ' seconds'));

        $latitude = 5352479;
        $longitude = 10000000;
        $elevation = $call % 1000;

        $trackPointMock = $this->createMock(TrackPoint::class);
        $trackPointMock
            ->expects(static::any())
            ->method('dateTime')
            ->will(static::returnValue($dateTime));
        $trackPointMock
            ->expects(static::any())
            ->method('latitude')
            ->will(static::returnValue(($latitude + $call) / 100000));
        $trackPointMock
            ->expects(static::any())
            ->method('longitude')
            ->will(static::returnValue(($longitude + $call) / 1000000));
        $trackPointMock
            ->expects(static::any())
            ->method('elevation')
            ->will(static::returnValue($elevation));
        if ($distance !== null) {
            $trackPointMock->expects(static::any())->method('hasDistance')->will(static::returnValue(true));
            $trackPointMock->expects(static::any())->method('distance')->will(static::returnValue($distance));
        }
        $trackPointMock->expects(static::any())->method('hasExtension')->with('HR')->will(static::returnValue(false));

        $call++;

        return $trackPointMock;
    }
}


tests/API/Authentication/AuthenticationTest.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo\Test\API\Authentication;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use SportTrackerConnector\Core\Tracker\Exception\InvalidCredentialsException;
use SportTrackerConnector\Endomondo\API\Authentication;

/**
 * Test case for Authentication.
 */
class AuthenticationTest extends \PHPUnit_Framework_TestCase
{

    public function testAuthenticationFromToken()
    {
        $token = 'my_token';
        $authentication = Authentication::fromToken($token);

        self::assertSame($token, $authentication->token());
        self::assertSame($token, (string)$authentication);
    }

    public function testAuthenticationFromUsernameAndPasswordThrowsClientExceptionOnBadResponse()
    {
        $mockHandler = new MockHandler(
            [
                new ClientException('Client error', $this->createMock(Request::class)),
            ]
        );

        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);

        $this->expectException(ClientException::class);

        Authentication::fromUsernameAndPassword('username', 'password', $client);
    }

    public function testAuthenticationFromUsernameAndPasswordThrowsInvalidCredentialsExceptionOnBadCredentials()
    {
        $mockHandler = new MockHandler(
            [
                new Response(200, [], 'USER_UNKNOWN')
            ]
        );

        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);

        $this->expectException(InvalidCredentialsException::class);

        Authentication::fromUsernameAndPassword('username', 'password', $client);
    }

    public function testAuthenticationFromUsernameAndPassword()
    {
        $token = 'my_token';

        $mockHandler = new MockHandler(
            [
                new Response(200, [],
                    sprintf(
                        "OK\naction=PAIRED\nauthToken=%s\nmeasure=METRIC\ndisplayName=Dragos\nuserId=123\nsecureToken=st",
                        $token
                    )
                )
            ]
        );

        $handler = HandlerStack::create($mockHandler);
        $client = new Client(['handler' => $handler]);
        $authentication = Authentication::fromUsernameAndPassword('username', 'password', $client);

        self::assertSame($token, $authentication->token());
        self::assertSame($token, (string)$authentication);
    }
}


tests/EndomondoTracker/EndomondoTrackerTest.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo\Test\EndomondoTracker;

use SportTrackerConnector\Core\Tracker\TrackerListWorkoutsResult;
use SportTrackerConnector\Core\Workout\Extension\HR;
use SportTrackerConnector\Core\Workout\SportMapperInterface;
use SportTrackerConnector\Core\Workout\Track;
use SportTrackerConnector\Core\Workout\Workout;
use SportTrackerConnector\Endomondo\API\Workouts;
use SportTrackerConnector\Endomondo\EndomondoTracker;
use SportTrackerConnector\Endomondo\SportMapper;

/**
 * Test for the Endomondo tracker.
 */
class EndomondoTrackerTest extends \PHPUnit_Framework_TestCase
{
    public function testId()
    {
        static::assertSame('endomondo', EndomondoTracker::ID());
    }

    public function testDownloadWorkoutSuccess()
    {
        $workoutId = '123';
        $json = \GuzzleHttp\json_decode(
            file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.json'),
            true
        );

        $workouts = $this->createMock(Workouts::class);
        $workouts
            ->expects(static::once())
            ->method('getWorkout')
            ->with($workoutId)
            ->will(self::returnValue($json));

        $endomondoTracker = new EndomondoTracker($workouts);
        $workout = $endomondoTracker->workout($workoutId);

        static::assertCount(1, $workout->tracks());

        $track = $workout->tracks()[0];
        static::assertSame(SportMapperInterface::CYCLING_SPORT, $track->sport());
        static::assertSame(128.045487, $track->length());
        static::assertEquals(new \DateTime('2014-06-04T18:05:32+0000'), $track->startDateTime());
        static::assertEquals(new \DateTime('2014-06-04T18:05:33+0000'), $track->endDateTime());
        static::assertSame(1, $track->duration()->totalSeconds());
        static::assertCount(2, $track->trackPoints());

        static::assertSame(129, $track->trackPoints()[0]->extension(HR::ID())->value());
        static::assertSame(130, $track->trackPoints()[1]->extension(HR::ID())->value());
    }

    public function testListWorkoutsWithEmptyList()
    {
        $startDate = new \DateTime('2016-01-01');
        $endDate = new \DateTime('2016-01-31');

        $workouts = $this->createMock(Workouts::class);
        $workouts
            ->expects(static::once())
            ->method('listWorkouts')
            ->with($startDate, $endDate)
            ->will(self::returnValue(array()));

        $endomondoTracker = new EndomondoTracker($workouts);
        $list = $endomondoTracker->workouts($startDate, $endDate);

        static::assertEmpty($list);
    }

    public function testListWorkoutsSuccess()
    {
        $startDate = new \DateTime('2016-01-01');
        $endDate = new \DateTime('2016-01-31');
        $json = \GuzzleHttp\json_decode(
            file_get_contents(__DIR__ . '/Fixtures/' . $this->getName() . '.json'),
            true
        );

        $workouts = $this->createMock(Workouts::class);
        $workouts
            ->expects(static::once())
            ->method('listWorkouts')
            ->with($startDate, $endDate)
            ->will(self::returnValue($json));

        $endomondoTracker = new EndomondoTracker($workouts);
        $list = $endomondoTracker->workouts($startDate, $endDate);

        static::assertEquals(
            array(
                new TrackerListWorkoutsResult(
                    '111111',
                    SportMapperInterface::RUNNING, new \DateTime('2014-07-24T18:45:00+0000')
                ),
                new TrackerListWorkoutsResult('222222',
                    SportMapperInterface::RUNNING,
                    new \DateTime('2014-07-24T16:55:00+0000')
                ),
                new TrackerListWorkoutsResult('333333',
                    SportMapperInterface::RUNNING,
                    new \DateTime('2014-07-22T18:32:00+0000')
                ),
            ),
            $list
        );
    }

    public function testUploadWorkout()
    {

        $workout = new Workout();
        $track1 = new Track([], SportMapperInterface::RUNNING);
        $track2 = new Track([], SportMapperInterface::SWIMMING);
        $workout->addTrack($track1);
        $workout->addTrack($track2);

        $workouts = $this->createMock(Workouts::class);
        $workouts
            ->expects(static::at(0))
            ->method('postTrack')
            ->with($track1, SportMapper::SPORT_RUNNING)
            ->will(self::returnValue('1111111'));
        $workouts
            ->expects(static::at(1))
            ->method('postTrack')
            ->with($track2, SportMapper::SPORT_SWIMMING)
            ->will(self::returnValue('2222222'));
        $endomondoTracker = new EndomondoTracker($workouts);
        $post = $endomondoTracker->post($workout);

        static::assertTrue($post);
    }
}


src/API/Workouts.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo\API;

use GuzzleHttp\Client;
use SportTrackerConnector\Core\Workout\Extension\HR;
use SportTrackerConnector\Core\Workout\Track;
use SportTrackerConnector\Core\Workout\TrackPoint;
use SportTrackerConnector\Endomondo\API\Exception\BadResponseException;

/**
 * Class for working with Endomondo API.
 */
class Workouts
{
    const URL_BASE = 'https://api.mobile.endomondo.com/mobile';
    const URL_WORKOUTS = 'https://api.mobile.endomondo.com/mobile/api/workouts';
    const URL_WORKOUT_GET = 'https://api.mobile.endomondo.com/mobile/api/workout/get';
    const URL_WORKOUT_POST = 'https://api.mobile.endomondo.com/mobile/api/workout/post';
    const URL_TRACK = 'https://api.mobile.endomondo.com/mobile/track';
    const URL_FRIENDS = 'https://api.mobile.endomondo.com/mobile/friends';

    const INSTRUCTION_PAUSE = 0;
    const INSTRUCTION_RESUME = 1;
    const INSTRUCTION_START = 2;
    const INSTRUCTION_STOP = 3;
    const INSTRUCTION_NONE = 4;
    const INSTRUCTION_GPS_OFF = 5;
    const INSTRUCTION_LAP = 6;

    /**
     * Endomondo authentication token.
     *
     * @var string
     */
    protected $authentication;

    /**
     * @var Client
     */
    protected $client;

    /**
     * @param Authentication $authentication
     * @param Client $client
     */
    public function __construct(Authentication $authentication, Client $client)
    {
        $this->authentication = $authentication;
        $this->client = $client;
    }

    /**
     * Get the details of a workout.
     *
     * Possible fields when getting the workout:
     *  device,simple,basic,motivation,interval,hr_zones,weather,polyline_encoded_small,points,lcp_count,tagged_users,pictures,feed
     *
     * @param string $idWorkout The ID of the workout.
     * @return array
     * @throws \RuntimeException
     */
    public function getWorkout(string $idWorkout) : array
    {
        $response = $this
            ->client
            ->get(
                self::URL_WORKOUT_GET,
                array(
                    'query' => array(
                        'authToken' => $this->authentication->token(),
                        'fields' => 'device,simple,basic,motivation,interval,weather,polyline_encoded_small,points,lcp_count,tagged_users,pictures',
                        'workoutId' => $idWorkout
                    )
                )
            );


        $json = \GuzzleHttp\json_decode($response->getBody(), true);
        if (isset($json['error'])) {
            throw new BadResponseException('Endomondo returned an unexpected error :"' . json_encode($json['error']));
        }

        return $json;
    }

    /**
     * Get a list of workouts in a date interval.
     *
     * @param \DateTime $startDate The start date for the workouts.
     * @param \DateTime $endDate The end date for the workouts.
     * @return array
     */
    public function listWorkouts(\DateTime $startDate, \DateTime $endDate) : array
    {
        $response = $this
            ->client
            ->get(
                self::URL_WORKOUTS,
                array(
                    'query' => array(
                        'authToken' => $this->authentication->token(),
                        'fields' => 'simple',
                        'maxResults' => 100000, // Be lazy and fetch everything in one request.
                        'after' => $startDate->format('Y-m-d H:i:s \U\T\C'),
                        'before' => $endDate->format('Y-m-d H:i:s \U\T\C')
                    )
                )
            );

        $json = \GuzzleHttp\json_decode($response->getBody(), true);

        return $json['data'];
    }

    /**
     * Post one workout track to Endomondo.
     *
     * @param Track $track
     * @param string $sport
     * @return null|string
     */
    public function postTrack(Track $track, string $sport)
    {
        $deviceWorkoutId = $this->generateDeviceWorkoutId();
        $duration = $track->duration()->totalSeconds();

        $workoutId = null;
        $previousPoint = null;
        $distance = 0;
        $speed = 0;
        // Split in chunks of 100 points like the mobile app.
        foreach (array_chunk($track->trackPoints(), 100) as $trackPoints) {
            $data = array();
            /** @var TrackPoint[] $trackPoints */
            foreach ($trackPoints as $trackPoint) {
                if ($trackPoint->hasDistance() === true) {
                    $distance = $trackPoint->distance();
                } elseif ($previousPoint !== null) {
                    $distance += $trackPoint->distanceFromPoint($previousPoint);
                }
                if ($previousPoint !== null) {
                    $speed = $trackPoint->speed($previousPoint);
                }

                $data[] = $this->flattenTrackPoint($trackPoint, $distance, $speed);

                $previousPoint = $trackPoint;
            }

            $this->postWorkoutData($deviceWorkoutId, $sport, $duration, $data);
        }

        // End of workout data.
        $data = $this->flattenEndWorkoutTrackPoint($track, $speed);
        $workoutId = $this->postWorkoutData($deviceWorkoutId, $sport, $duration, array($data));

        return $workoutId;
    }

    /**
     * Post the workout end data.
     *
     * @param Track $track The track.
     * @param float $speed The speed for the last point.
     * @return string The workout ID.
     */
    private function flattenEndWorkoutTrackPoint(Track $track, $speed)
    {
        $endDateTime = clone $track->endDateTime();
        $endDateTime->setTimezone(new \DateTimeZone('UTC'));
        $distance = $track->length();
        $lastTrackPoint = $track->lastTrackPoint();

        $totalAscent = $lastTrackPoint->elevation(); // TODO Compute it from the track, this is not correct.

        return $this->formatEndomondoTrackPoint(
            $endDateTime,
            self::INSTRUCTION_STOP,
            $lastTrackPoint->latitude(),
            $lastTrackPoint->longitude(),
            $distance,
            $speed,
            $totalAscent,
            $lastTrackPoint->hasExtension(HR::ID()) ? $lastTrackPoint->extension(HR::ID())->value() : ''
        );
    }

    /**
     * Post workout data chunk.
     *
     * @param string $deviceWorkoutId The workout ID in progress of the device.
     * @param string $sport The sport.
     * @param integer $duration The duration in seconds.
     * @param array $data The data points to post.
     * @return string The workout ID.
     * @throws \RuntimeException
     */
    private function postWorkoutData($deviceWorkoutId, $sport, $duration, array $data) : string
    {
        $body = \GuzzleHttp\Psr7\stream_for(gzencode(implode("\n", $data)));

        $response = $this
            ->client
            ->post(
                self::URL_TRACK,
                array(
                    'query' => array(
                        'authToken' => $this->authentication->token(),
                        'gzip' => 'true',
                        'workoutId' => $deviceWorkoutId,
                        'sport' => $sport,
                        'duration' => $duration,
                        'audioMessage' => 'false',
                        'goalType' => 'BASIC',
                        'extendedResponse' => 'true'
                    ),
                    'headers' => array(
                        'Content-Type' => 'application/octet-stream'
                    ),
                    'body' => $body
                )
            );

        $responseBody = $response->getBody()->getContents();
        $response = parse_ini_string($responseBody);

        if (array_key_exists('workout.id', $response)) {
            return $response['workout.id'];
        }

        throw new \RuntimeException('Unexpected response from Endomondo. Data may be partially uploaded. Response was: ' . $responseBody);
    }

    /**
     * Flatten a track point to be posted on Endomondo.
     *
     * @param TrackPoint $trackPoint The track point to flatten.
     * @param float $distance The total distance the point in meters.
     * @param float $speed The speed the point in km/h from the previous point.
     * @return string
     */
    private function flattenTrackPoint(TrackPoint $trackPoint, $distance, $speed) : string
    {
        $dateTime = clone $trackPoint->dateTime();
        $dateTime->setTimezone(new \DateTimeZone('UTC'));

        return $this->formatEndomondoTrackPoint(
            $dateTime,
            self::INSTRUCTION_START,
            $trackPoint->latitude(),
            $trackPoint->longitude(),
            $distance,
            $speed,
            $trackPoint->elevation(),
            $trackPoint->hasExtension(HR::ID()) ? $trackPoint->extension(HR::ID())->value() : ''
        );
    }

    /**
     * Format a point to send to Endomondo when posting a new workout.
     *
     * Type:
     *  0 - pause
     *  1 - running ?
     *  2 - running
     *  3 - stop
     *
     * @param \DateTime $dateTime
     * @param integer $type The post type (0-6). Don't know what they mean.
     * @param string $lat The latitude of the point.
     * @param string $lon The longitude of the point.
     * @param string $distance The distance in meters.
     * @param string $speed The speed in km/h.
     * @param string $elevation The elevation
     * @param string $heartRate The heart rate.
     * @param string $cadence The cadence (in rpm).
     * @return string
     */
    private function formatEndomondoTrackPoint(
        \DateTime $dateTime,
        $type,
        $lat = null,
        $lon = null,
        $distance = null,
        $speed = null,
        $elevation = null,
        $heartRate = null,
        $cadence = null
    ) : string
    {
        $dateTime = clone $dateTime;
        $dateTime->setTimezone(new \DateTimeZone('UTC'));

        return sprintf(
            '%s;%s;%s;%s;%s;%s;%s;%s;%s;',
            $dateTime->format('Y-m-d H:i:s \U\T\C'),
            $type,
            $lat,
            $lon,
            $distance / 1000,
            $speed,
            $elevation,
            $heartRate,
            $cadence
        );
    }

    /**
     * Generate a big number of specified length.
     *
     * @return string
     */
    private function generateDeviceWorkoutId()
    {
        $randNumber = '-';

        for ($i = 0; $i < 19; $i++) {
            $randNumber .= random_int(0, 9);
        }

        return $randNumber;
    }
}


src/EndomondoTracker.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo;

use DateTime;
use DateTimeZone;
use SportTrackerConnector\Core\Tracker\AbstractTracker;
use SportTrackerConnector\Core\Tracker\TrackerListWorkoutsResult;
use SportTrackerConnector\Core\Workout\Extension\HR;
use SportTrackerConnector\Core\Workout\SportMapperInterface;
use SportTrackerConnector\Core\Workout\Track;
use SportTrackerConnector\Core\Workout\TrackPoint;
use SportTrackerConnector\Core\Workout\Workout;
use SportTrackerConnector\Endomondo\API\Workouts;

/**
 * Endomondo tracker.
 */
class EndomondoTracker extends AbstractTracker
{
    /**
     * The Endomondo Workouts API.
     *
     * @var Workouts
     */
    protected $endomondoWorkouts;

    /**
     * @param Workouts $endomondoWorkouts
     */
    public function __construct(Workouts $endomondoWorkouts)
    {
        $this->endomondoWorkouts = $endomondoWorkouts;
    }

    /**
     * {@inheritdoc}
     */
    public function workout($idWorkout) : Workout
    {
        $json = $this->endomondoWorkouts->getWorkout($idWorkout);

        $workout = new Workout();
        $sport = $this->sportMapper()->sportFromCode((string)$json['sport']);
        $track = new Track(array(), $sport);

        if (array_key_exists('points', $json)) {
            foreach ($json['points'] as $point) {
                $trackPoint = new TrackPoint(
                    $point['lat'],
                    $point['lng'],
                    new DateTime($point['time'])
                );
                if (array_key_exists('alt', $point)) {
                    $trackPoint->setElevation($point['alt']);
                }
                if (array_key_exists('hr', $point)) {
                    $trackPoint->addExtension(new HR($point['hr']));
                }

                $track->addTrackPoint($trackPoint);
            }
        }

        $workout->addTrack($track);

        return $workout;
    }

    /**
     * {@inheritdoc}
     */
    public function workouts(DateTime $startDate, DateTime $endDate) : array
    {
        $list = array();
        $data = $this->endomondoWorkouts->listWorkouts($startDate, $endDate);
        foreach ($data as $workout) {
            $startDateTime = DateTime::createFromFormat(
                'Y-m-d H:i:s \U\T\C',
                $workout['start_time'],
                new DateTimeZone('UTC')
            );
            $list[] = new TrackerListWorkoutsResult(
                (string)$workout['id'],
                $this->sportMapper()->sportFromCode((string)$workout['sport']),
                $startDateTime
            );
        }

        return $list;
    }

    /**
     * {@inheritdoc}
     */
    public function post(Workout $workout) : bool
    {
        $workoutIds = array();
        foreach ($workout->tracks() as $track) {
            $sport = $this->sportMapper()->codeFromSport($track->sport());

            $workoutIds[] = $this->endomondoWorkouts->postTrack($track, $sport);
        }

        return count(array_filter($workoutIds)) > 0;
    }

    /**
     * {@inheritdoc}
     */
    protected function constructSportMapper() : SportMapperInterface
    {
        return new SportMapper();
    }

    /**
     * {@inheritdoc}
     */
    public static function ID() : string
    {
        return 'endomondo';
    }
}


src/API/Authentication.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo\API;

use GuzzleHttp\Client;
use Ramsey\Uuid\Uuid;
use SportTrackerConnector\Core\Tracker\Exception\InvalidCredentialsException;

class Authentication
{
    const URL_AUTHENTICATE = 'https://api.mobile.endomondo.com/mobile/auth';

    /**
     * @var string
     */
    protected $token;

    /**
     * @param string $token The token.
     */
    private function __construct(string $token)
    {
        $this->token = $token;
    }

    /**
     * Get the token.
     *
     * @return string
     */
    public function token() : string
    {
        return $this->token;
    }

    public static function fromToken(string $token) : Authentication
    {
        return new static($token);
    }

    /**
     * @param string $username
     * @param string $password
     * @param Client $client
     * @return Authentication If there is an error making the request.
     * @throws \SportTrackerConnector\Core\Tracker\Exception\InvalidCredentialsException
     */
    public static function fromUsernameAndPassword(string $username, string $password, Client $client) : Authentication
    {
        $response = $client->get(
            self::URL_AUTHENTICATE,
            array(
                'query' => array(
                    'country' => 'GB',
                    'action' => 'pair',
                    'deviceId' => (string)Uuid::uuid5(Uuid::NAMESPACE_DNS, gethostname()),
                    'email' => $username,
                    'password' => $password,
                )
            )
        );

        $responseBody = $response->getBody()->getContents();
        $response = parse_ini_string($responseBody);

        if (array_key_exists('authToken', $response)) {
            return new static($response['authToken']);
        }

        throw new InvalidCredentialsException('Authentication on Endomondo failed.');
    }

    public function __toString()
    {
        return $this->token();
    }
}


src/SportMapper.php
<?php

declare(strict_types = 1);

namespace SportTrackerConnector\Endomondo;

use SportTrackerConnector\Core\Workout\AbstractSportMapper;

/**
 * Sport mapper for Endomondo tracker.
 */
class SportMapper extends AbstractSportMapper
{
    const SPORT_RUNNING = '0';
    const SPORT_RUNNING_TREADMILL = '88';
    const SPORT_CYCLING_TRANSPORT = '1';
    const SPORT_CYCLING_SPORT = '2';
    const SPORT_MOUNTAIN_BIKING = '3';
    const SPORT_SKATING = '4';
    const SPORT_ROLLER_SKIING = '5';
    const SPORT_SKIING_CROSS_COUNTRY = '6';
    const SPORT_SKIING_DOWNHILL = '7';
    const SPORT_SNOWBOARDING = '8';
    const SPORT_KAYAKING = '9';
    const SPORT_KITE_SURFING = '10';
    const SPORT_ROWING = '11';
    const SPORT_SAILING = '12';
    const SPORT_WINDSURFING = '13';
    const SPORT_FITNESS_WALKING = '14';
    const SPORT_GOLF = '15';
    const SPORT_HIKING = '16';
    const SPORT_ORIENTEERING = '17';
    const SPORT_WALKING = '18';
    const SPORT_RIDING = '19';
    const SPORT_SWIMMING = '20';
    const SPORT_CYCLING_INDOOR = '21';
    const SPORT_OTHER = '22';
    const SPORT_AEROBICS = '23';
    const SPORT_BADMINTON = '24';
    const SPORT_BASEBALL = '25';
    const SPORT_BASKETBALL = '26';
    const SPORT_BOXING = '27';
    const SPORT_CLIMBING_STAIRS = '28';
    const SPORT_CRICKET = '29';
    const SPORT_CROSS_TRAINING = '30';
    const SPORT_DANCING = '31';
    const SPORT_FENCING = '32';
    const SPORT_FOOTBALL_AMERICAN = '33';
    const SPORT_FOOTBALL_RUGBY = '34';
    const SPORT_FOOTBALL_SOCCER = '35';
    const SPORT_HANDBALL = '36';
    const SPORT_HOCKEY = '37';
    const SPORT_PILATES = '38';
    const SPORT_POLO = '39';
    const SPORT_SCUBA_DIVING = '40';
    const SPORT_SQUASH = '41';
    const SPORT_TABLE_TENNIS = '42';
    const SPORT_TENNIS = '43';
    const SPORT_VOLLEYBALL_BEACH = '44';
    const SPORT_VOLLEYBALL_INDOOR = '45';
    const SPORT_WEIGHT_TRAINING = '46';
    const SPORT_YOGA = '47';
    const SPORT_MARTIAL_ARTS = '48';
    const SPORT_GYMNASTICS = '49';
    const SPORT_STEP_COUNTER = '50';
    const SPORT_CIRCUIT_TRAINING = '87';
    const SPORT_CLIMBING = '93';
    const SPORT_ICE_SKATING = '100';

    /**
     * {@inheritdoc}
     */
    public function getMap() : array
    {
        return array(
            self::RUNNING => self::SPORT_RUNNING,
            self::RUNNING_TREADMILL => self::SPORT_RUNNING_TREADMILL,
            self::WALKING => self::SPORT_WALKING,
            self::WALKING_FITNESS => self::SPORT_FITNESS_WALKING,
            self::CYCLING_SPORT => self::SPORT_CYCLING_SPORT,
            self::CYCLING_TRANSPORT => self::SPORT_CYCLING_TRANSPORT,
            self::CYCLING_INDOOR => self::SPORT_CYCLING_INDOOR,
            self::CYCLING_MOUNTAIN => self::SPORT_MOUNTAIN_BIKING,
            self::SWIMMING => self::SPORT_SWIMMING,
            self::GOLF => self::SPORT_GOLF,
            self::KAYAKING => self::SPORT_KAYAKING,
            self::KITE_SURFING => self::SPORT_KITE_SURFING,
            self::HIKING => self::SPORT_HIKING,
            self::SKATING => self::SPORT_SKATING,
            self::WEIGHT_TRAINING => self::SPORT_WEIGHT_TRAINING,
            self::OTHER => self::SPORT_OTHER,
        );
    }
}
